src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -4333,15 +4343,21 @@ void StorageReplicatedMergeTree::removePartAndEnqueueFetch(const String & part_n
             ops.emplace_back(zkutil::makeCheckRequest(fs::path(zookeeper_path) / "log", merge_pred.getVersion()));
         }
 
+        if (cancel_fetch_for_broken)
+        {
+            outdate_broken_part();
+            LOG_TRACE(log, "Cancel fetch for broken part {}", part_name);
```

Comment:
I'm sorry but I still do not understand what is the purpose of canceling the fetches.
As I mentioned above It is important that the broken part is labels as `part is lost forever`. Because it literally is lost forever.
That status is assigned when the replica is trying fetch a part, but it has not found on any replicas, for example when all replicas had detached it.
If I understand the diff correctly, your version just stops to fetch broken and detached part, that prevents clickhouse to mark that part as lost forever.

Reply:
I am ok that you have made broken parts are hardlinked in detached when it had failed to be copied to the detached directory.
But I think that we have to keep trying fetching it until other replicas have it. And when other replicas decided to detach it, we should get the situation when that part is marked as lost forever.